clang/lib/Sema/SemaDecl.cpp

```diff
@@ -8001,6 +8001,12 @@ NamedDecl *Sema::ActOnVariableDeclarator(
     }
   }
 
+  if (getLangOpts().HLSL) {
+    if (R->isHLSLSpecificType() && !NewVD->isImplicit()) {
+      Diag(D.getBeginLoc(), diag::err_hlsl_intangible_type_cannot_be_declared);
```

Comment:
Is the intent here to basically say that you can't create these types unless the declaration is implicit? I know we don't expose handle types in DXC, but do we really need to force these to be invalid in source?

Reply:
Yes, the intention was that these types can be created only in the implicit header `hlsl.h`. If that's not the case, I will remove this.

Reply:
`NewVD->isImplicit()` means the declaration is implicit and doesn't have any source, that prevents us from being able to use these in `hlsl.h` or directly in tests. I think that restriction may be more than we need.

Reply:
Got it, I thought `isImplicit()` also means declarations from implicit headers.

If we want to use these in `hlsl.h` then I don't think they can be truly sizeless. Sizeless types does not seem to be allowed in struct fields.

Reply:
HLSL's sizzles types will need to be allowed in struct fields, otherwise the property can't propagate correctly up to the resource objects or structures that contain resources which HLSL allows.